lib/centoswebpanel_installer.php

```diff
@@ -0,0 +1,203 @@
+        ];
+        $autoLoginUrl = 'https://' . $meta->host_name . ':' . $meta->port . '/v1/user_session';
+        $autoLoginRaw = $this->post($autoLoginData, $autoLoginUrl, 'POST');
+        $autoLoginResponse = json_decode($autoLoginRaw);
+        if ($autoLoginResponse == null || !isset($autoLoginResponse->msj->details[0]->token)) {
+            return;
+        }
+        $token = $autoLoginResponse->msj->details[0]->token;
+
+        if (strtolower($autoLoginResponse->status) == 'error') {
+            $this->Input->setErrors([
+                'login' => [
+                    'invalid' => 'Could not login to the remote server. Error : ' . $autoLoginResponse->msj
+                ]
+            ]);
+            return;
+        }
+
+        $loginData = ['username' => $serviceFields['centoswebpanel_username'], 'token' => $token];
+        $loginRaw = $this->post(
+            $loginData,
+            'https://' . $meta->host_name . ':2083/' . $serviceFields['centoswebpanel_username'] . '/',
+            'POST'
+        );
+        $loginResponse = json_decode($loginRaw);
+        if ($loginResponse == null) {
+            $this->Input->setErrors([
+                'login' => [
+                    'invalid' => 'Could not login to the remote server.'
+                ]
+            ]);
+            return;
+        }
+
+        // Make the Login system
+        $data = [
+            'softdomain' => $serviceFields['centoswebpanel_domain'],
+            // OPTIONAL - By default it will be installed in the /public_html folder
+            'softdirectory' => (!empty($configOptions['directory']) ? $configOptions['directory'] : ''),
+            'admin_username' => $configOptions['admin_name'],
+            'admin_pass' => $configOptions['admin_pass'],
+            'admin_email' => $client->email
+        ];
+
+        // List of Scripts
+        $scripts = $this->softaculousScripts();
+        $installationScript = $configOptions['script'];
+
+        // Which Script are we to install ?
+        foreach ($scripts as $key => $value) {
+            if (trim(strtolower($value['name'])) == trim(strtolower($installationScript))) {
+                $sid = $key;
+                break;
+            }
+        }
+
+        // Did we find the Script ?
+        if (empty($sid)) {
+            $this->Input->setErrors([
+                'script_id' => [
+                    'invalid' => 'Could not determine the script to be installed.'
+                    . ' Please make sure the script name is correct. Script Name : ' . $installationScript
+                ]
+            ]);
+            return;
+        }
+
+        // Install the script
+        $response = $this->scriptInstallRequest(
+            $sid,
+            $loginResponse->redirect_url,
+            $data
+        );
+
+
+        if ('installed' == strtolower($response)) {
+            return true;
+        } else {
+            $messages = unserialize($response);
```

Comment:
Is this something you need to unserialize? We really shouldn't be passing any user input to `unserialize` for reasons described @ https://www.php.net/manual/en/function.unserialize.php

Reply:
It is not user input, it is the response from the web panel and it is returned and serialized data.

Reply:
FYI the API supports different response formats, in this case JSON would be preferred
https://www.softaculous.com/docs/api/remote-api/#admin-functions

Also, anything that comes from somewhere else should be considered user input and not be trusted unless verified by some method.